File: pset6/credit/credit.py
from cs50 import get_string


def main():
    # Input card number from user
    while True:
        card_number = get_string("Number: ")
        if int(card_number) > 0:
            break

    # Check if the card number is valid
    if checksum(card_number) == True:
        print_card_company(card_number)
    else:
        print("INVALID")


# Check if the card number is valid
def checksum(card_number):
    # Find digits
    digits = len(card_number)

    # Convert str into int
    card_number = int(card_number)

    sum = 0
    for i in range(digits):
        # Find the number of digits
        digit = card_number % 10

        if i % 2 == 0:
            sum += digit
        else:
            sum += (digit * 2) // 10 + (digit * 2) % 10

        card_number //= 10

    # Check if the sum is divisible by 10
    if sum % 10 == 0:
        return True
    else:
        return False


def print_card_company(card_number):
    # Find the number of digits
    digits = len(card_number)

    if digits == 15 and (card_number[0:2] == "34" or card_number[0:2] == "37"):
        print("AMEX")
    elif digits == 16 and (card_number[0:2] == "51" or card_number[0:2] == "52" or card_number[0:2] == "53" or card_number[0:2] == "54" or card_number[0:2] == "55"):
        print("MASTERCARD")
    elif (digits == 13 or digits == 16) and card_number[0] == "4":
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

bool checksum(long n);
void print_card_company(long n);

int main(void)
{
    long card_number;
    do
    {
        card_number = get_long("Number:");
    }
    while (card_number < 0);

    if (checksum(card_number))
    {
        print_card_company(card_number);
    }
    else
    {
        printf("INVALID\n");
    }

}

bool checksum(long n)
{
    int sum = 0;
    for (int i = 0; n != 0; i++, n /= 10)
    {
        if (i % 2 == 0)
        {
            sum += n % 10;
        }

        else
        {
            int digit = n % 10;
            sum += (2 * digit) / 10 + (2 * digit) % 10;
        }
    }

    return (sum % 10) == 0;
}

void print_card_company(long n)
{

    if ((n >= 34e+13 && n < 35e+13) || (n >= 37e+13 && n < 38e+13))
    {
        printf("AMEX\n");
    }

    else if (n >= 51e+14 && n < 56e+14)
    {
        printf("MASTERCARD\n");
    }
    else if ((n >= 4e+12 && n < 5e+12) || (n >= 4e+15 && n < 5e+15))
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }
}